core/src/main/java/hudson/model/User.java

```diff
@@ -318,10 +318,26 @@ public synchronized void addProperty(@Nonnull UserProperty p) throws IOException
      * @since 1.419
      */
     public @Nonnull Authentication impersonate() throws UsernameNotFoundException {
+        return this.impersonate(this.getUserDetailsForImpersonation());
+    }
+    
+    /**
+     * This method checks with {@link SecurityRealm} if the user is a valid user that can login to the security realm.
+     * If {@link SecurityRealm} is a kind that does not support querying information about other users, this will
+     * use {@link LastGrantedAuthoritiesProperty} to pick up the granted authorities as of the last time the user has
+     * logged in.
+     *
+     * @throws UsernameNotFoundException
+     *      If this user is not a valid user in the backend {@link SecurityRealm}.
+     */
+    @Restricted(NoExternalUse.class)
+    public @CheckForNull UserDetails getUserDetailsForImpersonation() throws UsernameNotFoundException {
```

Comment:
This method and the following (#impersonate(UserDetails)) results of a split of the #impersonate() method in order to have access to the UserDetails. At the moment this is only used by BasicHeaderApiTokenAuthenticator to be able to trigger the event of SecurityListener without changing the API interface.